File: python/gsm_module.py
"""GSM module for sending SMS via serial communication."""
import asyncio
import logging
import serial
import serial.tools.list_ports
from typing import Optional, List, Dict, Any
from config import Config
from at_command_queue import ATCommandQueue, CommandObject

logger = logging.getLogger(__name__)


class GSMModule:
    """Manages GSM module communication via serial port."""
    
    def __init__(self, config: Config):
        self.config = config
        self.port: Optional[serial.Serial] = None
        self.response_buffer: str = ""
        self.command_queue: ATCommandQueue = ATCommandQueue(
            config.at_command_timeout, config.at_command_retry
        )
        self.is_ready: bool = False
        self.pending_command: Optional[CommandObject] = None
        self.pending_future: Optional[asyncio.Future] = None
        
        # Set executor for command queue
        self.command_queue.set_executor(self._execute_at_command)
    
    async def initialize(self) -> bool:
        """Initialize the GSM module."""
        logger.info("[GSM] Initializing GSM module...")
        
        try:
            serial_config = self.config.get_serial_config()
            self.port = serial.Serial(
                port=serial_config["path"],
                baudrate=serial_config["baudrate"],
                timeout=1,
            )
            
            # Start response reader task
            asyncio.create_task(self._read_responses())
            
            logger.info("[GSM] Serial port opened successfully")
            
            await asyncio.sleep(2)
            
            await self._send_command("AT", "OK")
            await self._send_command("ATE0", "OK")
            await self._send_command("AT+CMGF=1", "OK")
            await self._send_command("AT+CNMI=1,2,0,0,0", "OK")
            
            await self._check_network_registration()
            
            self.is_ready = True
            logger.info("[GSM] GSM module initialized successfully")
            
            return True
        except Exception as error:
            error_message = str(error)
            logger.error(f"[GSM] Initialization failed: {error_message}")
            raise
    
    async def _read_responses(self) -> None:
        """Continuously read responses from serial port."""
        while self.port and self.port.is_open:
            try:
                if self.port.in_waiting > 0:
                    data = self.port.readline().decode("utf-8", errors="ignore")
                    data = data.strip()
                    if data:
                        await self._handle_response(data)
                await asyncio.sleep(0.1)
            except Exception as e:
                logger.error(f"[GSM] Error reading response: {e}")
                await asyncio.sleep(0.5)
    
    async def _handle_response(self, data: str) -> None:
        """Handle incoming response data."""
        trimmed_data = data.strip()
        
        if not trimmed_data:
            return
        
        logger.info(f"[GSM] << {trimmed_data}")
        
        self.response_buffer += trimmed_data + "\n"
        
        if self.pending_command and self.pending_future:
            expected_response = self.pending_command.expected_response
            
            if "ERROR" in trimmed_data or "FAIL" in trimmed_data:
                self.pending_command = None
                if not self.pending_future.done():
                    self.pending_future.set_exception(Exception(trimmed_data))
                self.pending_future = None
                self.response_buffer = ""
                return
            
            if expected_response in trimmed_data or trimmed_data == expected_response:
                response = self.response_buffer
                self.response_buffer = ""
                self.pending_command = None
                if not self.pending_future.done():
                    self.pending_future.set_result(response)
                self.pending_future = None
                return
            
            if expected_response == ">" and trimmed_data == ">":
                self.response_buffer = ""
                self.pending_command = None
                if not self.pending_future.done():
                    self.pending_future.set_result(">")
                self.pending_future = None
                return
    
    async def _send_command(
        self, command: str, expected_response: str = "OK"
    ) -> str:
        """Send a command via the command queue."""
        if not self.port or not self.port.is_open:
            raise RuntimeError("Serial port not open")
        
        return await self.command_queue.add(command, expected_response)
    
    async def _execute_at_command(self, command_obj: CommandObject) -> str:
        """Execute an AT command and wait for response."""
        loop = asyncio.get_event_loop()
        future = loop.create_future()
        
        command = command_obj.command
        expected_response = command_obj.expected_response
        
        logger.info(f"[GSM] >> {command}")
        
        self.response_buffer = ""
        self.pending_command = command_obj
        self.pending_future = future
        
        # Set timeout
        timeout_handle = asyncio.create_task(
            self._command_timeout(command, future)
        )
        
        try:
            # Write command to serial port
            self.port.write((command + "\r\n").encode())
            
            # Wait for response
            result = await future
            timeout_handle.cancel()
            return result
        except Exception as e:
            timeout_handle.cancel()
            self.pending_command = None
            self.pending_future = None
            raise
    
    async def _command_timeout(
        self, command: str, future: asyncio.Future
    ) -> None:
        """Handle command timeout."""
        await asyncio.sleep(self.config.at_command_timeout / 1000.0)
        if not future.done():
            self.pending_command = None
            self.pending_future = None
            future.set_exception(
                TimeoutError(f"Command timeout: {command}")
            )
    
    async def send_sms(self, phone_number: str, message: str) -> bool:
        """Send an SMS message."""
        if not self.is_ready:
            raise RuntimeError("GSM module not ready")
        
        logger.info(f"[GSM] Sending SMS to {phone_number}: {message}")
        
        try:
            await self._send_command(f'AT+CMGS="{phone_number}"', ">")
            await asyncio.sleep(0.5)
            
            # Send message with Ctrl+Z terminator
            await self._execute_at_command(
                CommandObject(
                    command=message + chr(26),
                    expected_response="OK",
                )
            )
            
            logger.info(f"[GSM] SMS sent successfully to {phone_number}")
            return True
        except Exception as error:
            error_message = str(error)
            logger.error(
                f"[GSM] Failed to send SMS to {phone_number}: {error_message}"
            )
            raise
    
    async def send_alert(self, trigger_name: str) -> List[Dict[str, Any]]:
        """Send alert SMS to all configured phone numbers."""
        phone_numbers = self.config.get_phone_numbers()
        
        if not phone_numbers:
            return []
        
        from datetime import datetime
        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        message = f"ALERT: {trigger_name} triggered at {timestamp}"
        
        logger.info(f"[GSM] Sending alert: {message}")
        
        results = []
        
        for phone_number in phone_numbers:
            try:
                await self.send_sms(phone_number, message)
                results.append({"phoneNumber": phone_number, "success": True})
            except Exception as error:
                error_message = str(error)
                logger.error(
                    f"[GSM] Failed to send alert to {phone_number}: "
                    f"{error_message}"
                )
                results.append({
                    "phoneNumber": phone_number,
                    "success": False,
                    "error": error_message,
                })
        
        return results
    
    async def close(self) -> None:
        """Close the GSM module."""
        logger.info("[GSM] Closing GSM module...")
        
        self.command_queue.clear()
        
        if self.port and self.port.is_open:
            self.port.close()
        
        self.is_ready = False
        logger.info("[GSM] GSM module closed")
    
    async def _check_network_registration(self) -> None:
        """Check network registration status."""
        try:
            await self._send_command("AT+CREG?", "OK")
        except Exception:
            pass  # Ignore errors during registration check
    
    def get_status(self) -> Dict[str, Any]:
        """Get GSM module status."""
        return {
            "isReady": self.is_ready,
            "portOpen": self.port.is_open if self.port else False,
            "queueStatus": self.command_queue.get_status(),
        }

File: src/GSMModule.ts
import { SerialPort } from "serialport";
import { ReadlineParser } from "@serialport/parser-readline";
import { ATCommandQueue } from "./ATCommandQueue";
import { Config } from "./Config";
import { PendingCommand, SMSResult, GSMStatus, GSMDiagnostics } from "./types";
import {
  extractDiagnosticsFromResponse,
  getDetailedStatusReport,
  getCompactStatusReport,
} from "./utils/gsmDiagnosticsUtils";
import { logger, errorLogger } from "./utils/logger";

export { SMSResult, GSMStatus, GSMDiagnostics };

export class GSMModule {
  private config: Config;
  private port: SerialPort | null = null;
  private parser: ReadlineParser | null = null;
  private responseBuffer: string = "";
  private commandQueue: ATCommandQueue;
  private isReady: boolean = false;
  private pendingCommand: PendingCommand | null = null;
  private diagnostics: GSMDiagnostics = {};
  private isReconnecting: boolean = false;
  private reconnectAttempts: number = 0;
  private readonly MAX_RECONNECT_ATTEMPTS = 3;

  constructor(config: Config) {
    this.config = config;
    this.commandQueue = new ATCommandQueue(
      config.atCommandTimeout,
      config.atCommandRetry
    );

    this.commandQueue.setExecutor(async (commandObj) => {
      return await this.executeATCommand({
        command: commandObj.command,
        expectedResponse: commandObj.expectedResponse,
        skipCRLF: false,
      });
    });
  }

  public async performConnectionTest() {
    try {
      const tests = [
        {
          command: "AT+CPIN?",
          expectedResponse: "+CPIN:",
          log: "Checking PIN...",
        },
        {
          command: "AT+CMGF?",
          expectedResponse: "+CMGF:",
          log: "Checking mode...",
        },
        {
          command: "AT+CREG?",
          expectedResponse: "+CREG:",
          log: "Checking network registration...",
        },
        {
          command: "AT+CSQ",
          expectedResponse: "+CSQ:",
          log: "Checking signal quality...",
        },
        {
          command: "AT+CSCA?",
          expectedResponse: "+CSCA:",
          log: "Checking SC Address...",
        },
        {
          command: "AT+COPS?",
          expectedResponse: "+COPS:",
          log: "Checking current operator...",
        },
      ];

      for (const test of tests) {
        logger.info(`[GSM] ${test.log}`);
        const response = await this.sendCommand(
          test.command,
          test.expectedResponse
        );
        const diagnostics = extractDiagnosticsFromResponse(
          test.command,
          response
        );
        Object.assign(this.diagnostics, diagnostics);
      }
    } catch (error) {
      const errorMessage =
        error instanceof Error ? error.message : String(error);
      errorLogger.error("[GSM] Connection test failed:", errorMessage);
      throw error;
    }
  }

  public async initialize(): Promise<boolean> {
    logger.info("[GSM] Initializing GSM module...");

    try {
      this.port = new SerialPort({
        path: this.config.serialPort,
        baudRate: this.config.serialBaudrate,
        autoOpen: false,
      });

      this.parser = this.port.pipe(new ReadlineParser({ delimiter: "\r\n" }));

      this.parser.on("data", (data: string) => this.handleResponse(data));

      this.setupPortEventHandlers();

      await new Promise<void>((resolve, reject) => {
        this.port!.open((err) => {
          if (err) {
            reject(new Error(`Failed to open serial port: ${err.message}`));
          } else {
            resolve();
          }
        });
      });

      logger.info("[GSM] Serial port opened successfully");

      await this.delay(2000);

      await this.sendCommand("AT", "OK");
      await this.sendCommand("ATE0", "OK");
      await this.sendCommand("AT+CMGF=1", "OK");
      await this.sendCommand("AT+CNMI=1,2,0,0,0", "OK");
      await this.sendCommand(`AT+CSCS="GSM"`, "OK");
      await this.performConnectionTest();
      const statusReport = this.getDetailedStatusReport(this.config);
      logger.info(`\n${statusReport}\n`);
      logger.info(`[GSM] GSM module initialized successfully`);
      this.isReady = true;

      return true;
    } catch (error) {
      const errorMessage =
        error instanceof Error ? error.message : String(error);
      errorLogger.error("[GSM] Initialization failed:", errorMessage);
      throw error;
    }
  }

  private setupPortEventHandlers(): void {
    if (!this.port) return;

    this.port.on("close", () => {
      logger.warn("[GSM] Serial port closed unexpectedly");
      this.isReady = false;
      this.isReconnecting = false;
      this.reconnectAttempts = 0;

      if (this.pendingCommand) {
        clearTimeout(this.pendingCommand.timeoutHandle);
        this.pendingCommand.reject(
          new Error("Serial port closed during command execution")
        );
        this.pendingCommand = null;
      }
      this.commandQueue.clear();
    });

    this.port.on("error", (err: Error) => {
      errorLogger.error("[GSM] Serial port error:", err.message);
      this.isReady = false;

      if (this.pendingCommand) {
        clearTimeout(this.pendingCommand.timeoutHandle);
        this.pendingCommand.reject(
          new Error(`Serial port error: ${err.message}`)
        );
        this.pendingCommand = null;
      }
    });
  }

  private async ensurePortOpen(): Promise<void> {
    if (this.port && this.port.isOpen) {
      return;
    }

    if (this.isReconnecting) {
      let waitCount = 0;
      while (this.isReconnecting && waitCount < 50) {
        await this.delay(100);
        waitCount++;
        if (this.port && this.port.isOpen) {
          return;
        }
      }
      if (!this.port || !this.port.isOpen) {
        throw new Error("Serial port reconnection timeout");
      }
      return;
    }

    await this.reconnect();
  }

  private async reconnect(): Promise<void> {
    if (this.isReconnecting) {
      return;
    }

    this.isReconnecting = true;
    const initialAttempts = this.reconnectAttempts;

    while (
      this.reconnectAttempts - initialAttempts <
      this.MAX_RECONNECT_ATTEMPTS
    ) {
      this.reconnectAttempts++;
      const attemptNumber = this.reconnectAttempts - initialAttempts;

      logger.info(
        `[GSM] Attempting to reconnect serial port (attempt ${attemptNumber}/${this.MAX_RECONNECT_ATTEMPTS})...`
      );

      try {
        if (this.port) {
          try {
            if (this.port.isOpen) {
              await new Promise<void>((resolve) => {
                this.port!.close(() => resolve());
              });
            }
          } catch (error) {}
          this.port = null;
          this.parser = null;
        }

        this.port = new SerialPort({
          path: this.config.serialPort,
          baudRate: this.config.serialBaudrate,
          autoOpen: false,
        });

        this.parser = this.port.pipe(new ReadlineParser({ delimiter: "\r\n" }));
        this.parser.on("data", (data: string) => this.handleResponse(data));
        this.setupPortEventHandlers();

        await new Promise<void>((resolve, reject) => {
          this.port!.open((err) => {
            if (err) {
              reject(new Error(`Failed to reopen serial port: ${err.message}`));
            } else {
              resolve();
            }
          });
        });

        logger.info("[GSM] Serial port reconnected successfully");

        await this.delay(2000);
        await this.executeATCommandDirectly("AT", "OK");
        await this.executeATCommandDirectly("ATE0", "OK");
        await this.executeATCommandDirectly("AT+CMGF=1", "OK");
        await this.executeATCommandDirectly("AT+CNMI=1,2,0,0,0", "OK");
        await this.executeATCommandDirectly(`AT+CSCS="GSM"`, "OK");

        this.isReady = true;
        this.reconnectAttempts = 0;
        logger.info("[GSM] GSM module reinitialized after reconnection");
        this.isReconnecting = false;
        return;
      } catch (error) {
        const errorMessage =
          error instanceof Error ? error.message : String(error);
        errorLogger.error(
          `[GSM] Reconnection attempt ${attemptNumber} failed:`,
          errorMessage
        );

        if (attemptNumber >= this.MAX_RECONNECT_ATTEMPTS) {
          this.isReconnecting = false;
          this.reconnectAttempts = 0;
          throw new Error(
            `Failed to reconnect serial port after ${this.MAX_RECONNECT_ATTEMPTS} attempts`
          );
        }

        await this.delay(2000);
      }
    }

    this.isReconnecting = false;
    throw new Error(
      `Failed to reconnect serial port after ${this.MAX_RECONNECT_ATTEMPTS} attempts`
    );
  }

  private async executeATCommandDirectly(
    command: string,
    expectedResponse: string | null = "OK"
  ): Promise<string> {
    if (!this.port || !this.port.isOpen) {
      throw new Error("Serial port not open");
    }

    return new Promise((resolve, reject) => {
      if (command.includes(String.fromCharCode(26))) {
        logger.info(`[GSM] >> ${command.replace(/\x1A/g, "<CTRL+Z>")}`);
      } else {
        logger.info(`[GSM] >> ${command}`);
      }

      this.responseBuffer = "";

      const dataToWrite = command + "\r\n";

      // If expectedResponse is null, don't wait for response, resolve immediately
      if (expectedResponse === null) {
        this.port!.write(dataToWrite, (err) => {
          if (err) {
            reject(new Error(`Failed to write command: ${err.message}`));
          } else {
            resolve("");
          }
        });
        return;
      }

      const timeoutHandle = setTimeout(() => {
        this.pendingCommand = null;
        reject(new Error(`Command timeout: ${command}`));
      }, this.config.atCommandTimeout);

      this.pendingCommand = {
        command,
        expectedResponse,
        timeoutHandle,
        resolve,
        reject,
      };

      this.port!.write(dataToWrite, (err) => {
        if (err) {
          clearTimeout(timeoutHandle);
          this.pendingCommand = null;
          reject(new Error(`Failed to write command: ${err.message}`));
        }
      });
    });
  }

  private handleResponse(data: string): void {
    const trimmedData = data.trim();

    if (!trimmedData) return;

    logger.info(`[GSM] << ${trimmedData}`);

    this.responseBuffer += trimmedData + "\n";

    if (this.pendingCommand) {
      const { expectedResponse, timeoutHandle, resolve, reject } =
        this.pendingCommand;

      // If expectedResponse is null, we shouldn't have a pending command
      if (expectedResponse === null) {
        return;
      }

      if (trimmedData.includes("ERROR") || trimmedData.includes("FAIL")) {
        clearTimeout(timeoutHandle);
        this.pendingCommand = null;
        reject(new Error(trimmedData));
        return;
      }

      if (
        trimmedData.includes(expectedResponse) ||
        trimmedData === expectedResponse
      ) {
        clearTimeout(timeoutHandle);
        const response = this.responseBuffer;
        this.responseBuffer = "";
        this.pendingCommand = null;
        resolve(response);
        return;
      }

      if (expectedResponse === ">" && trimmedData === ">") {
        clearTimeout(timeoutHandle);
        this.pendingCommand = null;
        resolve(">");
        return;
      }
    }
  }

  public async sendCommand(
    command: string,
    expectedResponse: string | null = "OK"
  ): Promise<string> {
    await this.ensurePortOpen();
    if (!this.port || !this.port.isOpen) {
      return Promise.reject(new Error("Serial port not open"));
    }

    return this.commandQueue.add(command, expectedResponse);
  }

  private async executeATCommand(commandObj: {
    command: string;
    expectedResponse: string | null;
    skipCRLF?: boolean;
  }): Promise<string> {
    await this.ensurePortOpen();
    if (!this.port || !this.port.isOpen) {
      throw new Error("Serial port not open");
    }

    return new Promise((resolve, reject) => {
      const { command, expectedResponse, skipCRLF } = commandObj;

      if (skipCRLF && command.includes(String.fromCharCode(26))) {
        logger.info(`[GSM] >> ${command.replace(/\x1A/g, "<CTRL+Z>")}`);
      } else {
        logger.info(`[GSM] >> ${command}`);
      }

      this.responseBuffer = "";

      const dataToWrite = skipCRLF ? command : command + "\r\n";

      // If expectedResponse is null, don't wait for response, resolve immediately
      if (expectedResponse === null) {
        this.port!.write(dataToWrite, (err) => {
          if (err) {
            reject(new Error(`Failed to write command: ${err.message}`));
          } else {
            resolve("");
          }
        });
        return;
      }

      const timeoutHandle = setTimeout(() => {
        this.pendingCommand = null;
        reject(new Error(`Command timeout: ${command}`));
      }, this.config.atCommandTimeout);

      this.pendingCommand = {
        command,
        expectedResponse,
        timeoutHandle,
        resolve,
        reject,
      };

      this.port!.write(dataToWrite, (err) => {
        if (err) {
          clearTimeout(timeoutHandle);
          this.pendingCommand = null;
          reject(new Error(`Failed to write command: ${err.message}`));
        }
      });
    });
  }

  private async sendSMS(
    phoneNumber: string,
    message: string,
    performConnectionTest: boolean = true
  ): Promise<boolean> {
    if (!this.isReady) {
      throw new Error("GSM module not ready");
    }

    logger.info(`[GSM] Sending SMS to ${phoneNumber}: ${message}`);

    try {
      if (performConnectionTest) {
        await this.performConnectionTest();
      }
      logger.info("[GSM] Sending SMS...");
      await this.sendCommand(`AT+CMGS="${phoneNumber}"`, null);

      await this.delay(600);
      await this.executeATCommand({
        command: message + String.fromCharCode(26),
        expectedResponse: "OK",
        skipCRLF: true,
      });

      logger.info(`[GSM] SMS sent successfully to ${phoneNumber}`);
      return true;
    } catch (error) {
      const errorMessage =
        error instanceof Error ? error.message : String(error);
      errorLogger.error(
        `[GSM] Failed to send SMS to ${phoneNumber}:`,
        errorMessage
      );
      throw error;
    }
  }

  public async sendToAll(message: string): Promise<SMSResult[]> {
    const phoneNumbers = this.config.phoneNumbers;

    if (phoneNumbers.length === 0) {
      return [];
    }

    logger.info(`[GSM] Sending to all: ${message}`);
    const results: SMSResult[] = [];

    for (const phoneNumber of phoneNumbers) {
      try {
        await this.sendSMS(phoneNumber, message, false);
        results.push({ phoneNumber, success: true });
      } catch (error) {
        const errorMessage =
          error instanceof Error ? error.message : String(error);
        errorLogger.error(
          `[GSM] Failed to send to ${phoneNumber}:`,
          errorMessage
        );
        results.push({ phoneNumber, success: false, error: errorMessage });
      }
    }

    return results;
  }

  public async close(): Promise<void> {
    logger.info("[GSM] Closing GSM module...");

    this.commandQueue.clear();

    if (this.port && this.port.isOpen) {
      await new Promise<void>((resolve) => {
        this.port!.close((err) => {
          if (err) {
            errorLogger.error("[GSM] Error closing port:", err.message);
          }
          resolve();
        });
      });
    }

    this.isReady = false;
    logger.info("[GSM] GSM module closed");
  }

  private delay(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  public getStatus(): GSMStatus {
    return {
      isReady: this.isReady,
      portOpen: this.port ? this.port.isOpen : false,
      queueStatus: this.commandQueue.getStatus(),
    };
  }

  public getDiagnostics(): GSMDiagnostics {
    return { ...this.diagnostics };
  }

  public getDetailedStatusReport(config: Config): string {
    return getDetailedStatusReport(this.diagnostics, config);
  }

  public getCompactStatusReport(activeTriggers: Set<string>): string {
    const inputs = [
      activeTriggers.has("trigger1"),
      activeTriggers.has("trigger2"),
      activeTriggers.has("trigger3"),
    ];
    return getCompactStatusReport(this.diagnostics, this.config, inputs);
  }
}
